swarm/network/simulations/discovery/dynamic_test.go

```diff
@@ -0,0 +1,359 @@
+				return true, err
+			}
+		case err := <-subs[id].Err():
+			if err != nil {
+				log.Error(fmt.Sprintf("error getting peer events for node %v", id), "err", err)
+			}
+		default:
+		}
+		return false, err
+	}
+
+	timeout := 30 * time.Second
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	result := simulations.NewSimulation(net).Run(ctx, &simulations.Step{
+		Action:  action,
+		Trigger: trigger,
+		Expect: &simulations.Expectation{
+			Nodes: ids,
+			Check: check,
+		},
+	})
+	if result.Error != nil {
+		t.Fatal(result.Error)
+	}
+
+	wg := sync.WaitGroup{}
+	for i := 0; i < int(numUpDowns); i++ {
+		wg.Add(1)
+		dur := ctrl.randomDelay(0)
+		time.Sleep(dur)
+		go func() {
+			if err := ctrl.nodeUpDown(); err != nil {
+				wg.Done()
+				t.Fatal(err)
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+
+	log.Warn("exiting test")
+}
+
+type nodeCtrl struct {
+	seq                int
+	net                *simulations.Network
+	upNodes            []discover.NodeID
+	upAddrs            [][]byte
+	nodeCursor         int
+	pot                map[discover.NodeID]*network.PeerPot
+	mu                 sync.Mutex
+	healthCheckRetries int
+	healthCheckDelay   time.Duration
+}
+
+func newNodeCtrl(net *simulations.Network, nodes []discover.NodeID) *nodeCtrl {
+	ctrl := &nodeCtrl{
+		net:                net,
+		healthCheckRetries: defaultHealthCheckRetires,
+		healthCheckDelay:   defaultHealthCheckDelay,
+	}
+	for ctrl.nodeCursor = 0; ctrl.nodeCursor < len(nodes); ctrl.nodeCursor++ {
+		ctrl.upNodes = append(ctrl.upNodes, nodes[ctrl.nodeCursor])
+		ctrl.upAddrs = append(ctrl.upAddrs, network.ToOverlayAddr(ids[ctrl.nodeCursor].Bytes()))
+		log.Debug("init nodeCtrl", "idx", ctrl.nodeCursor, "upnode", ctrl.upNodes[ctrl.nodeCursor], "upaddr", fmt.Sprintf("%x", ctrl.upAddrs[ctrl.nodeCursor]))
+	}
+	ctrl.pot = network.NewPeerPot(testMinProxBinSize, ctrl.upNodes, ctrl.upAddrs)
+	for k, p := range ctrl.pot {
+		for i, nn := range p.NNSet {
+			log.Debug("init nodeCtrl nnset", "i", i, "node", k.TerminalString(), "nn", fmt.Sprintf("%x", nn))
+		}
+	}
+	return ctrl
+}
+
+func (self *nodeCtrl) nodeUpDown() error {
+	mu.Lock()
+	self.seq++
+	nodeIdx := rand.Intn(len(self.upNodes) - 1)
+
+	nodeId := self.upNodes[nodeIdx]
+	log.Info("Restart node: stop", "id", nodeId, "addr", fmt.Sprintf("%x", self.upAddrs[nodeIdx]), "upnodes", len(self.upNodes))
+	n := self.net.GetNode(nodeId)
+	self.upNodes[nodeIdx] = self.upNodes[len(self.upNodes)-1]
+	self.upAddrs[nodeIdx] = self.upAddrs[len(self.upAddrs)-1]
+	self.upNodes = self.upNodes[:len(self.upNodes)-1]
+	self.upAddrs = self.upAddrs[:len(self.upAddrs)-1]
+	self.nodeCursor--
+	self.pot = network.NewPeerPot(testMinProxBinSize, self.upNodes, self.upAddrs)
+	//	if err := self.net.Stop(nodeId); err != nil {
+	//		return err
+	//	}
+	for _, c := range self.net.Conns {
+		if (c.One == nodeId || c.Other == nodeId) && c.Up {
+			err := self.net.Disconnect(c.One, c.Other)
```

Comment:
Why is `net.Disconnect` better than `net.GetNode(id).Stop()`? No risk he same peer could be suggested/added again?

Reply:
Stop actually creates a _new_ kademlia object, which doesn't know about any peers. It's not the behavior we're testing. When the persistent peer store is merged, `simulations.Network.Stop()` can be used.